### PR TITLE
feat: Adds cachedFlatMap operator

### PR DESCRIPTION
**Pull request by mdboon**

### CODE REVIEW

**Review by andogq — APPROVED** [review]

**Comment by giraugh** [comment]
Would this work as a special case of like a `.flatReduce()` or something like that? Could then build a helper for working with a cache on top of that?

```ts
someStream
   .flatReduce($.cacheReducer(
     new Map(),
     atom => /* snip: determine key */,
     atom => /* snip: determine value */,
   ))
```

Would also leave the door open to other reductions like summing
```ts
someStream
  .flatReduce($.sumReducer(atom => produceAsyncNumber(atom)))
```

Or a manual setup
```ts
someStream
  .flatReduce((acc, atom) => doAsyncStuff(acc, atom)) // should return a stream that emits once w/ next acc val
```

File: src/stream/higher-order.ts [hunk 1]
```diff
@@ -125,6 +125,54 @@ export class HigherOrderStream<T, E> extends StreamTransforms<T, E> {
...
+     * @group Higher Order
+     */
+    cachedFlatMap<U>(
+        cb: (value: T) => MaybePromise<Stream<U, E>>,
+        keyFn: (value: T) => string | number | symbol,
+    ): Stream<U, E> {
+        const trace = this.trace("cachedFlatMap");
+
+        return this.consume(async function* (it) {
+            const cache = new Map<string | number | symbol, Atom<U, E>[]>();
```
Comment (giraugh) [inline src/stream/higher-order.ts]:
```suggestion
            const cache = new Map<PropertyKey, Atom<U, E>[]>();
```

**Comment by andogq** [comment]
> Would this work as a special case of like a `.flatReduce()` or something like that? Could then build a helper for working with a cache on top of that?
> 
> ```ts
> someStream
>    .flatReduce($.cacheReducer(
>      new Map(),
>      atom => /* snip: determine key */,
>      atom => /* snip: determine value */,
>    ))
> ```

I think it's worthwhile having a publicly exposed `cachedFlatMap` (as Mark's done here), but it would be good if it were implemented in terms of `flatReduce` under the hood. I don't think we should worry about it for now, given that `flatReduce` doesn't exist atm.

(also mark is getting antsy, please help)

**Comment by giraugh** [comment]
Sounds good to me 👍 
Roger roger, sending more ants 🐜 🐜 🐜 🐜

**Pull request merged**